lib/matplotlib/axis.py

```diff
@@ -1571,10 +1571,6 @@ def have_units(self):
         return self.converter is not None or self.units is not None
 
     def convert_units(self, x):
-        # If x is already a number, doesn't need converting
-        if munits.ConversionInterface.is_numlike(x):
```

Comment:
Is it really necessary this way? The purpose of converters is to convert to numbers. So the original logic "if is number return else convert" seems more clear than "convert; if that fails, did we already have a number? Ok then we just use it".

Maybe `munits.ConversionInterface.is_numlike()` is just not the right function/name. I'd prefer this to stay and be

~~~
if not munits.ConversionInterface.needs_conversion(x):
    return x
~~~

That will allow to cleanly pass through `Decimal`s.

Reply:
Well, not sure on the sign of the logic `is_natively_supported_numlike(x)` may be better than `not needs_conversion(x)`.

Reply:
The original cause of this problem is the operations like plus between Decimal and float is not supported by matplotlib. If there is not such operations in the logic, Decimal input will not cause any error.  As we discussed before, it doesn't quite make sense to distinguish Decimal and other types of number since Decimal is a type of Number.

Reply:
That's why I'm saying "is_numlike" is not the right check. Semantically, we want
~~~
if matplotlib_can_natively_work_with(x):
     return x
# try to convert x
~~~

It's not about number or not, it's about "do we need a converter". So, if matplotlib cannot natively handle `Decimal` that can fall through to the converter.

Reply:
How can we define natively handle? For Decimal input, matplotlib supports it if there is no operation between Decimal and float. For example:
```
x = [1, 2, 3, 4]
y = [Decimal(1.5), Decimal(2.5), Decimal(3.5), Decimal(4.5)]
ax.bar(x, y, align='center')
```
This piece of code works fine. I am not sure about other kinds of plot, but based on this, I think matplotlib does not totally support/unsupport Decimal.

Reply:
Well, operationally, "can natively handle" is currently `is_numlike(x) and not isinstance(x, Decimal)`. I mean, that's what your proposed code does as well.

It's a bit clumsy to define is that way, but that's the current logic. And I prefer to have that logic explicitly written down in a single place in a function.

Reply:
I see currently is_numlike is used in some places like StrCategoryConverter.convert() in category.py. And from the code (as below), it looks like the Decimal input will not cause any problems here (since Decimal will be converted to floats).
```
# pass through sequence of non binary numbers
if all((units.ConversionInterface.is_numlike(v) and
        not isinstance(v, (str, bytes))) for v in values):
    return np.asarray(values, dtype=float)
```
So I think maybe instead of change is_numlike(x), which cause API changes, we can add a new function is_natively_supported_numlike(x) like proposed in ConversionInterface. What do you think?

Reply:
I think it's a good idea to leave `is_numlike(x)` as is and introduce and use `ConversionInterface.is_natively_supported_numlike(x)`

*Naming*: Not sure if we need the `_numlike` suffix in the name. On the one hand, this is exactly meant as `is_natively_supported` - no artificial restriction to numlikes necessary (even though it's unlikely that we'll *natively* support non-numlikes). On the other hand, the `_numlike` may give additional context to the reader, but I would see this more as a job of the `is_natively_supported*` docstring.

`is_numlike(x)` may become deprecated in the future, but we can handle this later.

Reply:
I have added a function named is_natively_supported() in ConversionInterface. Since the current unit system assume lists are homogeneous, I followed the same style as is_numlik(). Now I only consider all numbers except Decimal are natively supported.